src/imap/mod.rs

```diff
@@ -532,98 +536,102 @@ impl Imap {
-
-        let config = &mut self.config;
-        let mailbox = config
-            .selected_mailbox
-            .as_ref()
-            .ok_or_else(|| format_err!("No mailbox selected, folder: {}", folder))?;
-
-        let new_uid_validity = match mailbox.uid_validity {
-            Some(v) => v,
-            None => {
-                bail!("No UIDVALIDITY for folder {:?}", folder);
-            }
-        };
-
-        if new_uid_validity == uid_validity {
-            return Ok((uid_validity, last_seen_uid));
+    ) -> Result<(u32, bool)> {
+        let newly_selected = self.select_folder(context, Some(folder)).await?;
+
+        let mailbox = &mut self.config.selected_mailbox.as_ref();
+        let mailbox =
+            mailbox.with_context(|| format!("No mailbox selected, folder: {}", folder))?;
+
+        let new_uid_validity = mailbox
+            .uid_validity
+            .with_context(|| format!("No UIDVALIDITY for folder {}", folder))?;
+
+        let old_uid_validity = get_uidvalidity(context, folder).await;
+        let old_uid_next = get_uid_next(context, folder).await;
+
+        if new_uid_validity == old_uid_validity {
+            let new_emails = if newly_selected == NewlySelected::No {
+                // The folder was not newly selected i.e. no SELECT command was run. This means that mailbox.uid_next
+                // was not updated and largest_uid may contain an incorrect value. So, just return true so that
+                // the caller tries to fetch new messages (we could of course run a SELECT command now, but trying to fetch
+                // new messages is only one command, just as a SELECT command)
+                true
+            } else if let Some(uid_next) = mailbox.uid_next {
+                uid_next != old_uid_next // If uid_next changed, there are new emails
+            } else {
+                true
+            };
+            return Ok((old_uid_next, new_emails));
         }
 
-        if mailbox.exists == 0 {
-            info!(context, "Folder \"{}\" is empty.", folder);
+        // // TODO can this be safely deleted?:
+        // if mailbox.exists == 0 {
+        //     info!(context, "Folder \"{}\" is empty.", folder);
 
-            // set lastseenuid=0 for empty folders.
-            // id we do not do this here, we'll miss the first message
-            // as we will get in here again and fetch from lastseenuid+1 then
+        //     // set lastseenuid=0 for empty folders.
+        //     // id we do not do this here, we'll miss the first message
+        //     // as we will get in here again and fetch from lastseenuid+1 then
 
-            set_config_last_seen_uid(context, &folder, new_uid_validity, 0).await;
-            return Ok((new_uid_validity, 0));
-        }
+        //     set_config_last_seen_uid(context, &folder, new_uid_validity, 0).await;
+        //     return Ok((0, false));
+        // }
```

Comment:
Can this be safely deleted? Probably yes because 
1. if we would miss the first message in a folder on every fetch, then the tests would probably fail
2. we are not fetching from lastseenuid+1 anymore but from uid_next (without +1)

But I thought I'd rather ask @dignifiedquire (you wrote this code?)